### DISCUSSION TITLE
p:print

### DISCUSSION THREAD

**Discussion by tony-in-nz** [primefaces/community#1909] [PrimeFaces] [answered]
I am using Atlantis Layout and did development work, but I found that I cannot use p:printer. When I invoke p:printer for browser preview, it displays a blank page. Additionally, when I use Ctrl+P, all pages appear blank. Why is this happening, and how can I resolve it?

The page is below (I stripped everything out to create this page).

Google search and forum search was not much help.

```
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"
 xmlns:h="http://java.sun.com/jsf/html"
 xmlns:f="http://java.sun.com/jsf/core"
 xmlns:ui="http://java.sun.com/jsf/facelets"
 xmlns:p="http://primefaces.org/ui">

<!-- USE .jsf extension when accessing the page directly -->
<!-- USE .jsf extension when accessing the page directly -->
<!-- USE .jsf extension when accessing the page directly -->

<h:head>
	<f:facet name="first">
		<meta http-equiv="X-UA-Compatible" content="IE=edge" />
		<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" />
		<meta name="apple-mobile-web-app-capable" content="yes" />
	</f:facet>

	<title>Print Test</title>

	<h:outputScript name="js/layout.js" library="atlantis-layout"/>
	<h:outputScript name="js/prism.js" library="atlantis-layout"/>

	<meta name="application-name" content="XXX" />
	<meta name="author" content="XXX" />
	<meta name="msapplication-TileColor" content="#FFFFFF" />

	<link rel="apple-touch-icon" sizes="180x180" href="#{request.contextPath}/xxx-layout/icons/apple-touch-icon.png"/>
	<link rel="icon" type="image/png" sizes="32x32" href="#{request.contextPath}/xxx-layout/icons/favicon-32x32.png"/>
	<link rel="icon" type="image/png" sizes="16x16" href="#{request.contextPath}/xxx-layout/icons/favicon-16x16.png"/>
	<link rel="manifest" href="#{request.contextPath}/xxx-layout/icons/site.webmanifest"/>

	<link href="#{request.contextPath}/xxx-layout/icons/favicon.ico" rel="shortcut icon" type="image/x-icon"/>
</h:head>
<h:body styleClass="main-body">

	<h:form id="printForm">
		<div>
			<p:commandButton id="btn" value="Print Text">
				<p:printer target="output" />
			</p:commandButton>
			<br/>
			<h:outputText id="output" value="PrimeFaces Rocks!" />
			<hr/>
			<p:commandButton id="lnk" value="Print Image">
				<p:printer target="image" />
			</p:commandButton>
			<hr/>
			<p:graphicImage id="image" name="/site-images/tony/tony-suit.jpg" library="xxx-layout" style="width:50%; display: block; margin-left: auto; margin-right: auto;" />
		</div>
	</h:form>

	<h:outputStylesheet name="css/primeicons.css" library="atlantis-layout"/>
	<h:outputStylesheet name="css/primeflex.min.css" library="atlantis-layout"/>
	<h:outputStylesheet name="css/layout-blue.css" library="atlantis-layout" />
	<h:outputStylesheet name="css/xxx.css" library="xxx-layout"/>
	<h:outputStylesheet name="primeicons/primeicons.css" library="primefaces" />
	
</h:body>
</html>
```

![repo](https://github.com/primefaces/community/assets/8173858/ee050c6a-d563-48ef-8cf0-e38e7decfd17)

clicking on Print Text gives:
![blank](https://github.com/primefaces/community/assets/8173858/b25a43d8-47f6-4f78-b979-f44d5b033ff1)

**Comment by melloware** [comment]
Have you tried a newer PrimeFaces than 12.0.7?  13.0.10? 14.0.0?

**Comment by tony-in-nz** [comment]
??

The latest in Elite is 12.0.7

So you are saying just use the comminity versions from now on?

![PFElite](https://github.com/primefaces/community/assets/8173858/05e76c69-b50a-45eb-ae1e-db5df84a8def)

**Comment by tony-in-nz** [comment]
I will use 13.0.1 from here

![PFCommunity](https://github.com/primefaces/community/assets/8173858/e80a4e71-ef65-47d3-81ad-60ab3cecfe9d)

**Comment by tony-in-nz** [comment]
As I am using java 21 (Quarkus) I think I still have to add the jakarta classifier...


```
		<!-- https://mvnrepository.com/artifact/org.primefaces/primefaces -->
		<!-- We use the community version unless the Elite version is higher version -->
		<dependency>
			<groupId>org.primefaces</groupId>
			<artifactId>primefaces</artifactId>
			<version>${primefaces.version}</version>
			<classifier>jakarta</classifier>
		</dependency>
```

this gives:
2024-05-15 12:26:30,376 INFO  [org.pri.web.PostConstructApplicationEventListener] (Quarkus Main Thread) Running on PrimeFaces 13.0.1 

Sadly the print screen is still empty.

I think I may have soemthing missing from the head section???

**Comment by tony-in-nz** [comment]
Tried 14.0 too
2024-05-15 12:34:48,545 INFO  [org.pri.web.PostConstructApplicationEventListener] (Quarkus Main Thread) Running on PrimeFaces 14.0.0                                                       

Sadly; no change.

**Comment by tony-in-nz** [comment]
FYI I am using the Quarkus bridge:
		<myfaces.version>4.0.2</myfaces.version>

In PF 14 I get a new error.

Cannot find a referenced form:

Inside a non nested for:
```
									<h:form id="tagsCloud">

										<div class="grid">
											<div class="col-12 md:col-12 lg:col-6">
												<p:tagCloud model="#{xxxController.skillsTagCloudModel}" style="width:100%; height:100%">
													<p:ajax event="select" update="messagesForm:growl" listener="#{xxxController.onSelect}" />
												</p:tagCloud>
											</div>

											<div class="col-12 md:col-12 lg:col-6">
												<p:tagCloud model="#{xxxController.industriesTagCloudModel}" style="width:100%; height:100%">
													<p:ajax event="select" update="messagesForm:growl" listener="#{xxxCVController.onSelect}" />
												</p:tagCloud>
											</div>
										</div>

									</h:form>
```

I have another form in another area of the page:

```
			<h:form id="messagesForm">
				<div id="messagesArea">
					<p:growl id="growl" globalOnly="true" showDetail="false">
						<p:autoUpdate />
					</p:growl>
				</div>
			</h:form>
```

and the error is when the page is shown:
Cannot find component for expression "messagesForm:growl" referenced from "tagsCloud:j_id_28".

This works fine in 13.

**Reply by melloware** [reply]
Yes you need to use the proper JSF expression in 14.0.0 which is ":messagesForm:growl".  Forms must start with a colon.

**Reply by tony-in-nz** [reply]
That fixed that referenced component (I love specifications; so many to choose from)....
Now back to the print problem :)

**Reply by melloware** [reply]
As for p:print can you check the PrimeFaces Showcase which is 13.0.10 and see if the Print Example is working?

Also if you are using Quarkus the Quarkus PrimeFaces extension already handles the PrimeFaces versions for you?

**Reply by melloware** [reply]
Here is QUARKUS faces deployed on PF 14.0.0. https://quarkus-faces-melloware-8a6a34c1.koyeb.app/ui/misc/printer.xhtml?jfwid=18c13

Does that URL work for you and does it print?

**Reply by tony-in-nz** [reply]
It does fill in the print dialog.

**Reply by tony-in-nz** [reply]
I removed PF14 from the POM and it was not there (you implied a PF would be auto added by  the Quarkus the Quarkus PrimeFaces extension.

**Comment by tony-in-nz** [comment]
As for the print now working in PF14:
This is shown in the console.
```
VM4363:1 Uncaught ReferenceError: $ is not defined
    at <anonymous>:1:1
    at r (printer.js.jsf?ln=primefaces&v=14.0.0:2:480)
    at u (printer.js.jsf?ln=primefaces&v=14.0.0:4:45)
    at d.print.d.fn.print (printer.js.jsf?ln=primefaces&v=14.0.0:8:85)
    at HTMLButtonElement.<anonymous> (tonyherstellcv.jsf:31:898)
    at Object.bcn (core.js.jsf?ln=primefaces&v=14.0.0:31:114)
    at HTMLButtonElement.onclick (tonyherstellcv.jsf:31:760)
```

**Reply by melloware** [reply]
That error '$' means JQuery is not loaded. I don't see how that is possible unless you have something whacky going on in your page?  Does my Quarkus Print link above work for you?

**Reply by melloware** [reply]
Also your CSS includes are all wrong aren't they?  You have PrimeIcons CSS twice and you have PrimeFlex coming from Altantis?

**Reply by tony-in-nz** [reply]
Stripped down page:

```
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"
 xmlns:h="http://java.sun.com/jsf/html"
 xmlns:f="http://java.sun.com/jsf/core"
 xmlns:ui="http://java.sun.com/jsf/facelets"
 xmlns:p="http://primefaces.org/ui">

<!-- USE .jsf extension when accessing the page directly -->
<!-- USE .jsf extension when accessing the page directly -->
<!-- USE .jsf extension when accessing the page directly -->

<h:head>
	<f:facet name="first">
		<meta http-equiv="X-UA-Compatible" content="IE=edge" />
		<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" />
		<meta name="apple-mobile-web-app-capable" content="yes" />
	</f:facet>

	<title>Print Test</title>

	<meta name="application-name" content="XXX" />
	<meta name="author" content="Tony Herstell" />
	<meta name="msapplication-TileColor" content="#FFFFFF" />

	<link rel="apple-touch-icon" sizes="180x180" href="#{request.contextPath}/xxx-layout/icons/apple-touch-icon.png"/>
	<link rel="icon" type="image/png" sizes="32x32" href="#{request.contextPath}/xxx-layout/icons/favicon-32x32.png"/>
	<link rel="icon" type="image/png" sizes="16x16" href="#{request.contextPath}/xxx-layout/icons/favicon-16x16.png"/>
	<link rel="manifest" href="#{request.contextPath}/xxx-layout/icons/site.webmanifest"/>

	<link href="#{request.contextPath}/xxx-layout/icons/favicon.ico" rel="shortcut icon" type="image/x-icon"/>
</h:head>
<h:body styleClass="main-body">

	<h:form id="printForm">
		<div>
			<p:commandButton id="btn" value="Print Text">
				<p:printer target="output" />
			</p:commandButton>
			<br/>
			<h:outputText id="output" value="PrimeFaces Rocks!" />
			<hr/>
			<p:commandButton id="lnk" value="Print Image">
				<p:printer target="image" />
			</p:commandButton>
			<hr/>
			<p:graphicImage id="image" name="/site-images/tony/tony-suit.jpg" library="xxx-layout" style="width:50%; display: block; margin-left: auto; margin-right: auto;" />
		</div>
	</h:form>

	<h:outputStylesheet name="css/layout-blue.css" library="atlantis-layout" />
	<h:outputStylesheet name="css/xxx.css" library="xxx-layout"/>
	<h:outputStylesheet name="primeicons/primeicons.css" library="primefaces" />
	
</h:body>
</html>
```

**Reply by melloware** [reply]
I will test it tomorrow. Basically you just want an image in the Print dialog right?

**Reply by melloware** [reply]
@tony-in-nz here is a simple reproducer just unzip and run `mvn quarkus:dev`.

[quarkus-print.zip](https://github.com/primefaces/community/files/15321313/quarkus-print.zip)


1. Notice I don't declare any PrimeFaces version yet because i am using `quarkus-primefaces` it automatically loads PrimeFaces 14.0.0 for me.

2. Notice that both Print buttons work fine.

This is as bare bones and simple of a Quarkus example as I can get.

**Reply by tony-in-nz** [reply]
I emailed you an updated zip of yours.
The printing to .pdf stops when I added in the theme (atlantis-layout).
I got the two sides of the page to suddenly work; but the print PDF stops working.
jar was too big to add to this message.

**Reply by melloware** [reply, marked as answer]
OK the issue is in the Atlantis layout in `invoice.scss` you need to remove this section it adds Media Print queries which break the p:printer.

```css
@media print {
    body * {
        visibility: hidden;
    }
    #invoice-content * {
        visibility: visible;
    }
    #invoice-content {
        width: 100%;
        position: absolute;
        left: 0;
        top: 0;
        padding: 0;
        margin: 0;
        background: #ffffff;
        color: #2d353c;
    }

    .invoice {
        .invoice-to {
            border-top: 1px solid #e3e3e3;
        }

        .invoice-items table tr {
            border-bottom: 1px solid #e3e3e3;
        }
    }
}
```

**Reply by melloware** [reply]
@tony-in-nz is it safe for me to close this issue did it resolve your issue?

**Reply by tony-in-nz** [reply]
Thank you SOOO much.
Commenting all this stuff out from all the files (quite a few) worked.
Please close (and possibly raise a bug on the themes to fix this for the
next release)?
Taking it all out is one thing; but it must have been there for a reason I
would assume.
Tony

**Reply by melloware** [reply]
It was only in their for the Invoice showcase Demo.
